src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java

```diff
@@ -0,0 +1,221 @@
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package engineering.swat.watch.impl.jdk;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import engineering.swat.watch.WatchEvent;
+import engineering.swat.watch.WatchScope;
+import engineering.swat.watch.impl.EventHandlingWatch;
+
+public class JDKFileTreeWatch extends JDKBaseWatch {
+    private final Logger logger = LogManager.getLogger();
+    private final Path rootPath;
+    private final Path relativePathParent;
+    private final Map<Path, JDKFileTreeWatch> childWatches = new ConcurrentHashMap<>();
+    private final JDKBaseWatch internal;
+
+    public JDKFileTreeWatch(Path fullPath, Executor exec,
+            BiConsumer<EventHandlingWatch, WatchEvent> eventHandler) {
+        this(fullPath, Path.of(""), exec, eventHandler);
+    }
+
+    public JDKFileTreeWatch(Path rootPath, Path relativePathParent, Executor exec,
+            BiConsumer<EventHandlingWatch, WatchEvent> eventHandler) {
+
+        super(rootPath.resolve(relativePathParent), exec, eventHandler);
+        this.rootPath = rootPath;
+        this.relativePathParent = relativePathParent;
+
+        var internalEventHandler = eventHandler.andThen(new ChildWatchesUpdater());
+        this.internal = new JDKDirectoryWatch(path, exec, internalEventHandler) {
+
+            // Override to ensure that this watch relativizes events wrt
+            // `rootPath` (instead of `path`, as is the default behavior)
+            @Override
+            public WatchEvent relativize(WatchEvent event) {
+                return new WatchEvent(event.getKind(), rootPath,
+                    rootPath.relativize(event.calculateFullPath()));
```

Comment:
I think we should be able to calculate the relative path based on where we are in the tree? instead of doing a `calculateFullPath`?